### PR TITLE
Added support for a per-projector cache, including an LRU implementation

### LINKED ISSUE #2: Support a projection cache
* Represented by a `IProjectionCache<TProjection>` with a default implementation based on https://github.com/dennisdoomen/fluidcaching
* A cache doesn't need to be thread-safe since it will only be accessed by a single thread at the time.
* Each child-projector should be able to use its own cache. 
* Who's responsible to reset the cache when a projector is being restarted for some reason? _The code that uses the `NHibernateProjector`_

### PR DESCRIPTION
**Pull request by dennisdoomen**
Closes #2 

* [x]  The `NHibernateProjector` will try to get the projection from the cache first. If it doesn't exist, it will load or create a projection and store it in the cache and the NH session. 
* [x]  If a projection is mapped to be deleted, the projector should evict the item from the cache.
* [x] ~~If a custom map is used, the custom implementation should get access to the cache to add, evict or update its data.~~ _Decided that this is the responsibility of the custom handler, which should already have access to the cache_
* [x] Add readme about custom changes to the session and effect on the cache
* [x] Add readme item about relationship and how the cache doesn't solve this for you
* [x] Clear the cache if any exception happens

### CODE REVIEW

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        void Add(TProjection projection);
+
+        /// <summary>
+        /// Attempts to get the item identified by <paramref name="key"/> from the cache, or creates a new one. 
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios. So multiple concurrent requests for the same key must always 
+        /// return the same object.
+        /// </remarks>
+        Task<TProjection> Get(TKey key, Func<Task<TProjection>> createProjection);
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/IProjectionCache.cs]: Why do we want to support concurrent access to the cache?

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    /// </summary>
+    public interface IProjectionCache<TProjection, in TKey>
+    {
+        /// <summary>
+        /// Adds a new item to the cache. 
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios.
+        /// </remarks>
+        void Add(TProjection projection);
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/IProjectionCache.cs]: What should happen if the projection with such key is already in the cache?

File: Src/LiquidProjections.NHibernate/LiquidProjections.NHibernate.csproj [hunk 3]
```diff
@@ -24,7 +24,7 @@
     <DocumentationFile>bin\Debug\LiquidProjections.NHibernate.XML</DocumentationFile>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <NoWarn>1571, 1591, 1570, 1573</NoWarn>
-    <LangVersion>6</LangVersion>
+    <LangVersion>default</LangVersion>
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/LiquidProjections.NHibernate.csproj]: I would avoid using `default` here. That means that any contributor can use the latest version of C# and force everyone else to upgrade.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        /// </remarks>
+        Task<TProjection> TryGet(TKey key);
+
+        /// <summary>
+        /// Removes the item identified by <paramref name="key"/> from the cache.
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios and be idempotent. 
+        /// </remarks>
+        void Remove(TKey key);
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/IProjectionCache.cs]: We could also cache the fact that a projection does not exist. But maybe that's a rare situation when we check whether the projection exists and don't create the projection afterwards.

File: Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs [hunk 5]
```diff
@@ -74,6 +81,11 @@ internal sealed class NHibernateEventMapConfigurator<TProjection, TKey>
             }
             else
             {
+                if (!context.Session.Contains(projection))
+                {
+                    context.Session.SaveOrUpdate(projection);
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs]: It's always an existing projections and it doesn't have any changes, right? So we have to neither save nor update it. We just want to attach it to the session so the future changes to it are tracked. `context.Session.Lock(projection, LockMode.None)` should do that.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace LiquidProjections.NHibernate
+{
+    /// <summary>
+    /// Defines a write-through cache that the <see cref="NHibernateProjector{TProjection,TKey,TState}"/> can use to avoid unnecessary loads
+    /// of the projection.
+    /// </summary>
+    public interface IProjectionCache<TProjection, in TKey>
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/IProjectionCache.cs:10]: Do we support caching for projections which have relations to other NHibernate entities? If not, we should document that limitation somewhere.

File: Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs [hunk 7]
```diff
@@ -74,6 +81,11 @@ internal sealed class NHibernateEventMapConfigurator<TProjection, TKey>
             }
             else
             {
+                if (!context.Session.Contains(projection))
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs]: What if somehow the projector has loaded the projection from the database without trying to get if from the cache first? In that case the session contains a different instance of the same projection.

File: Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs [hunk 8]
```diff
@@ -132,9 +145,8 @@ internal sealed class NHibernateEventMapConfigurator<TProjection, TKey>
             else
             {
                 context.Session.Delete(existingProjection);
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs:147]: Same here:

> What if somehow the projector has loaded the projection from the database without trying to get if from the cache first? In that case the session contains a different instance of the same projection.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+        /// <summary>
+        /// Attempts to get the item identified by <paramref name="key"/> from the cache.
+        /// Returns <c>null</c> if the item is not in the cache.
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios. So multiple concurrent requests for the same key must always 
+        /// return the same object.
+        /// </remarks>
+        Task<TProjection> TryGet(TKey key);
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/IProjectionCache.cs]: Why is this method asynchronous but `Add` and `Remove` are not?

If the cache access can be a lengthy operation in some implementation then `TryGet` should be asynchronous indeed. But in that case `Remove` at least should be asynchronous too as we want to make sure that the cache is not outdated. `Add` could still be fire-and-forget operation but better not as it should still respect the order of different operations even in that case.

If only the cache-miss handler is asynchronous then only `Get` should be asynchronous.

File: Tests/LiquidProjections.NHibernate.Specs/Properties/AssemblyInfo.cs [hunk 10]
```diff
@@ -31,8 +31,8 @@
...
 // by using the '*' as shown below:
-// [assembly: AssemblyVersion("0.1.0.0")]
-[assembly: AssemblyVersion("0.1.0.0")]
-[assembly: AssemblyFileVersion("0.1.0.0")]
+// [assembly: AssemblyVersion("2.0.1.0")]
+[assembly: AssemblyVersion("2.0.1.0")]
+[assembly: AssemblyFileVersion("2.0.1.0")]
 
-[assembly: AssemblyInformationalVersion("0.1.0+0.Branch.master.Sha.6117c2c4ea7d90855b3a2f961d7e2b2dbeb1e955")]
+[assembly: AssemblyInformationalVersion("2.0.1-Cache.1+1.Branch.Cache.Sha.0bc38fefb932210db2dfe42b49e4e79fe4ed2e69")]
```
Comment (IharBury) [inline Tests/LiquidProjections.NHibernate.Specs/Properties/AssemblyInfo.cs]: I don't think we should modify this file in every pull request

File: Src/LiquidProjections.NHibernate/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -30,7 +30,7 @@
 //
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
-// [assembly: AssemblyVersion("0.1.0.0")]
+// [assembly: AssemblyVersion("2.0.1.0")]
 
-[assembly: AssemblyInformationalVersion("0.1.0+0.Branch.master.Sha.6117c2c4ea7d90855b3a2f961d7e2b2dbeb1e955")]
-[assembly: AssemblyFileVersion("0.1.0.0")]
+[assembly: AssemblyInformationalVersion("2.0.1-Cache.1+1.Branch.Cache.Sha.0bc38fefb932210db2dfe42b49e4e79fe4ed2e69")]
+[assembly: AssemblyFileVersion("2.0.1.0")]
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/Properties/AssemblyInfo.cs]: I don't think we should modify this file in every pull request

File: Tests/LiquidProjections.NHibernate.Specs/packages.config [hunk 12]
```diff
@@ -1,14 +1,14 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Chill" version="3.0.0" targetFramework="net452" />
-  <package id="FluentAssertions" version="4.19.2" targetFramework="net452" />
+  <package id="Chill" version="3.0.1" targetFramework="net452" />
+  <package id="FluentAssertions" version="5.0.0-beta0001" targetFramework="net452" />
```
Comment (IharBury) [inline Tests/LiquidProjections.NHibernate.Specs/packages.config:4]: Should we depend on a beta?

File: Tests/LiquidProjections.NHibernate.Specs/packages.config [hunk 13]
```diff
@@ -1,14 +1,14 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
```
Comment (IharBury) [inline Tests/LiquidProjections.NHibernate.Specs/packages.config:2]: Would be better to update packages in separate commit(s)

**Review by IharBury — CHANGES_REQUESTED** [review]

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs (hunk 6 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/IProjectionCache.cs:10]: I expect NHibernate to figure this out for me. You would essentially be passing the entire graph into a new session. NH will use the version/identity to figure out what happened.

File: Tests/LiquidProjections.NHibernate.Specs/packages.config (hunk 12 above)
Comment (dennisdoomen) [reply Tests/LiquidProjections.NHibernate.Specs/packages.config:4]: It's an official beta which will never go away, so I don't see a problem.

File: Tests/LiquidProjections.NHibernate.Specs/packages.config (hunk 13 above)
Comment (dennisdoomen) [reply Tests/LiquidProjections.NHibernate.Specs/packages.config]: Yeah, I agree. But it was too late.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs (hunk 2 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/IProjectionCache.cs]: Documented this.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/IProjectionCache.cs]: On second thought, I don't think that is necessary. We can't guarantee that everything runs from the same thread, but at least we don't access the thread concurrently.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs (hunk 9 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/IProjectionCache.cs]: The only reason why `Get` is asynchronous is because the code behind the `createProjection` delegate might be I/O intensive. I've removed `TryGet`.

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs (hunk 4 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/IProjectionCache.cs]: > But maybe that's a rare situation when we check whether the projection exists and don't create the projection afterwards.

Why would we not create/load the projection if it doesn't exist in the cache?

File: Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs (hunk 7 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs]: That would mean somebody tinkered with the session directly without us knowing about it. How do you suggest we handle that? Update the cache with the existing item? Or overwrite the session regardless?

File: Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs (hunk 8 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs]: Do we care?

File: Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs (hunk 5 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs]: What made you think that method can do that?

File: Src/LiquidProjections.NHibernate/IProjectionCache.cs (hunk 6 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.NHibernate/IProjectionCache.cs:10]: Discussed this f2f and agree to add a limitation to the readme.

**Comment by dennisdoomen** [comment]
All comments addressed.

File: Src/LiquidProjections.NHibernate/PassthroughCache.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+            // Do nothing.
+        }
+
+        public void Clear()
+        {
+            // Do nothing
+        }
+
+        public Task<TProjection> TryGet(TKey key)
```
Comment (IharBury) [inline Src/LiquidProjections.NHibernate/PassthroughCache.cs:48]: Not used

**Review by IharBury — APPROVED** [review]
